R/teal_lockfile.R

```diff
@@ -0,0 +1,114 @@
+#' Generate lockfile for application reproducibility
+#'
+#' This function is invoked during [teal::init] to create `renv`-compatible lockfile for use within the application.
+#'
+#' The function leverages [renv::snapshot()], which offers multiple methods for lockfile creation.
+#'
+#' - User-specified:
+#'     - **Pre-computed lockfile**: Users can provide their own pre-computed lockfile by specifying the path via
+#'     `teal.renv.lockfile` option. Automatic lockfile computation is skipped in such case.
+#' - Automatically computed:
+#'     - **Working directory lockfile**: If `teal.renv.lockfile` is not set, `teal` will, by default, create an
+#'      `implicit` type lockfile that uses `renv::dependencies()` to detect all R packages in the current project's
+#'      working directory.
+#'     - **`DESCRIPTION`-based lockfile**: To generate a lockfile based on a `DESCRIPTION` file in your working
+#'     directory, set `renv::settings$snapshot.type("explicit")`. The naming convention for `type` follows
+#'     `renv::snapshot()`. For the `"explicit"` type, refer to `renv::settings$package.dependency.fields()` for the
+#'     `DESCRIPTION` fields included in the lockfile.
+#'     - **Custom files-based lockfile**: To specify custom files as the basis for the lockfile, set
+#'     `renv::settings$snapshot.type("custom")` and configure the `renv.snapshot.filter` option.
+#'
+#' @section lockfile usage:
+#' After creating the lockfile, you can restore the application environment using `renv::restore()`.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()].
+#'
+#' @return Nothing. This function is executed for its side effect of creating a lockfile used in the `teal` application.
+#'
+#' @keywords internal
+teal_lockfile <- function() {
+  lockfile_path <- "teal_app.lock"
+  # If user has setup the file, there is no need to compute a new one.
+  user_lockfile <- getOption("teal.renv.lockfile", "")
+  if (!identical(user_lockfile, "")) {
+    if (file.exists(user_lockfile)) {
+      file.copy(user_lockfile, lockfile_path)
+      return(invisible(NULL))
+    } else {
+      stop("lockfile provided through options('teal.renv.lockfile') does not exist.")
+    }
+  }
+
+  if (!(is_in_test() || is_r_cmd_check())) {
+    old_plan <- future::plan()
+    # If there is already a parallel (non-sequential) backend, reuse it.
+    if (inherits(old_plan, "sequential")) {
+      future::plan(future::multisession, workers = 2)
```

Comment:
Why `workers = 2` while it is a single renv task? 
Can it work on a single core machines? 
Why not to use `plan(multisession)`?

Reply:
If workers = 1, then there is just one worker and can handle one process at a time. If there are two workers then one handles shiny session and the other one handles the task that runs `renv::restore()`. It's not `additional workers`, but `workers all together used by R session`.


You can set more workers than there is cores on a machine. This will yield a warning that it is ineffective, but it will not stop.

```r
> future::availableCores()
system 
    12 
> future::plan(future::multisession, workers = future::availableCores()+1)
Warning message:
In checkNumberOfLocalWorkers(workers) :
  Careful, you are setting up 13 localhost parallel workers with only 12 CPU cores available for this R process (per ‘system’), which could result in a 108% load. The soft limit is set to 100%. Overusing the CPUs has negative impact on the current R process, but also on all other processes of yours and others running on the same machine. See help("parallelly.options", package = "parallelly") for how to override the soft and hard limits
> future::plan() -> current_plan
> current_plan
multisession:
- args: function (..., workers = c(system = 13), envir = parent.frame())
- tweaked: TRUE
- call: future::plan(future::multisession, workers = future::availableCores() + 1)
```

Reply:
I wonder then, that happens if somebody runs `teal::init` on a single-core machine?

Reply:
I think we can skip the discussion about single core processors.

Reply:
On a single core machine it will allow you to set two workers, but that does not matter and the execution will be done sequentially anyway if there is only 1 worker on machine. This means that `teal::init()` will not finish until `teal::teal_lockfile()` is done.